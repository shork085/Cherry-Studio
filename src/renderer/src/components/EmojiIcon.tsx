import { FC } from 'react'
import styled from 'styled-components'

interface EmojiIconProps {
  emoji: string
  className?: string
  size?: number
  fontSize?: number
}

const EmojiIcon: FC<EmojiIconProps> = ({ emoji, className, size = 26, fontSize = 15 }) => {
  return (
    <Container className={className} $size={size} $fontSize={fontSize}>
      <EmojiBackground>{emoji || '⭐️'}</EmojiBackground>
      {emoji}
    </Container>
  )
}

const Container = styled.div<{ $size: number; $fontSize: number }>`
  width: ${({ $size }) => $size}px;
  height: ${({ $size }) => $size}px;
  border-radius: ${({ $size }) => $size / 2}px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  font-size: ${({ $fontSize }) => $fontSize}px;
  position: relative;
  overflow: hidden;
  margin-right: 3px;
`

const EmojiBackground = styled.div`
  width: 100%;
  height: 100%;
  position: absolute;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 200%;
  transform: scale(1.5);
  filter: blur(5px);
  opacity: 0.4;
`

export default EmojiIcon
