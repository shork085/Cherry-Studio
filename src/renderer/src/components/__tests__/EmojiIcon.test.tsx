import { render } from '@testing-library/react'
import { describe, expect, it } from 'vitest'

import EmojiIcon from '../EmojiIcon'

describe('EmojiIcon', () => {
  it('should render with provided emoji', () => {
    const { container } = render(<EmojiIcon emoji="üöÄ" />)

    // Should render the emoji
    expect(container.textContent).toContain('üöÄ')

    // Should also render emoji in background
    const background = container.querySelector('div > div')
    expect(background?.textContent).toContain('üöÄ')
  })

  it('should render default emoji when no emoji provided', () => {
    const { container } = render(<EmojiIcon emoji="" />)

    // Background should have default star emoji
    const background = container.querySelector('div > div')
    expect(background?.textContent).toContain('‚≠êÔ∏è')

    // Foreground should be empty (the actual emoji prop value)
    const emojiContainer = container.firstChild as HTMLElement
    // Remove background text to get only foreground text
    const foregroundText = emojiContainer.textContent?.replace(background?.textContent || '', '')
    expect(foregroundText).toBe('')
  })

  it('should apply custom className', () => {
    const customClass = 'custom-emoji-class'
    const { container } = render(<EmojiIcon emoji="üòä" className={customClass} />)

    const emojiContainer = container.firstChild as HTMLElement
    expect(emojiContainer).toHaveClass(customClass)
  })

  it('should match snapshot', () => {
    const { container } = render(<EmojiIcon emoji="üéâ" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should handle special emojis correctly', () => {
    const specialEmojis = ['üë®‚Äçüíª', 'üèÉ‚Äç‚ôÄÔ∏è', 'üë®‚Äçüë©‚Äçüëß‚Äçüë¶', 'üá®üá≥']

    specialEmojis.forEach((emoji) => {
      const { container } = render(<EmojiIcon emoji={emoji} />)
      expect(container.textContent).toContain(emoji)
    })
  })

  it('should apply custom size and fontSize props', () => {
    const { container } = render(<EmojiIcon emoji="üåü" size={40} fontSize={24} />)
    const emojiContainer = container.firstChild as HTMLElement

    // Verify that the component renders with custom props
    expect(emojiContainer).toHaveStyle({ width: '40px', height: '40px' })
    expect(emojiContainer).toHaveStyle({ fontSize: '24px' })
  })

  it('should handle empty string emoji', () => {
    const { container } = render(<EmojiIcon emoji="" />)
    const backgroundElement = container.querySelector('div > div')

    // Should show default emoji in background when emoji is empty
    expect(backgroundElement?.textContent).toContain('‚≠êÔ∏è')
  })
})
