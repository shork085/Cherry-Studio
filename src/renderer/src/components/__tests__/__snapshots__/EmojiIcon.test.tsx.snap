// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmojiIcon > should match snapshot 1`] = `
.c0 {
  width: 26px;
  height: 26px;
  border-radius: 13px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  font-size: 15px;
  position: relative;
  overflow: hidden;
  margin-right: 3px;
}

.c1 {
  width: 100%;
  height: 100%;
  position: absolute;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 200%;
  transform: scale(1.5);
  filter: blur(5px);
  opacity: 0.4;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    ðŸŽ‰
  </div>
  ðŸŽ‰
</div>
`;
